syncer/peers/heap.go

```diff
@@ -0,0 +1,114 @@
+package peers
+
+import (
+	"container/heap"
+	"sync"
+)
+
+type Peer interface {
+	// identifier
+	ID() string
+	// comparator
+	Less(Peer) bool
+}
+
+type PeerHeap struct {
+	sync.RWMutex
+	peers     []Peer
+	lookupMap map[string]int
+}
+
+func NewPeerHeap(peers []Peer) *PeerHeap {
+	peerHeap := &PeerHeap{
+		peers:     peers,
+		lookupMap: make(map[string]int, len(peers)),
+	}
+
+	heap.Init(peerHeap)
+
+	for i, p := range peerHeap.peers {
+		peerHeap.lookupMap[p.ID()] = i
+	}
+
+	return peerHeap
+}
+
+// Put puts peer into heap
+// Case 1: Appends a peer if it doesn't exist
+// Case 2: Update peer info and reorder if it exists already
+func (h *PeerHeap) Put(peer Peer) {
+	h.RWMutex.Lock()
+	defer h.RWMutex.Unlock()
+
+	index, ok := h.lookupMap[peer.ID()]
+	if ok {
+		// exists already
+		h.peers[index] = peer
+		heap.Fix(h, index)
```

Comment:
Why are you re-adding a peer if they are already in the heap? What kind of information can change that would trigger a reorder?

Reply:
(Latest Block Height, Distance) is used for ordering heap. Latest Block Height is not constant value, we need to update this when receiving new status

Reply:
@zivkovicmilos